alphagov/paas-cf
# [#146074923] Monitor CF API availability during deployment

## What

We want to monitor the availability of the CF API during a deploy.

As a first step we do not want to test _everything_ and this change focuses on monitoring a couple of read-only API calls, but with a view to enabling adding more checks in the future.

To achieve this goal this change:

* Adds a new `api-availability-tests` job to the pipeline designed to start before the `cf-deploy` job begins and end after `cf-deploy` finishes.
* Adds a testing/monitoring system that repeatably tries blocks of CF API calls and monitors the number of successes and failures and reports back any collected errors during the run.
* Adds two relatively simple blocks of CF API calls to the monitor the equivalent of `cf apps` and `cf app myapp`

## Notes

The general idea here is similar to the app-availability-tests that use the vegeta library to put an app under load during deploy and monitor the response, only rather than focusing on single HTTP requests we are interested in sets of CF API calls that simulate users doing actions with the CF CLI. 

## How to review

* [x] Read/Review the individual commits
* [x] Deploy to a dev environment
    * check that the new test job runs during cf-deploy.
    * check that the new test job has some useful output at the end.
    * try and force a failure (by for example removing the `API-AVAILABILITY-*` app deployed to the healthchecks space during a deploy).
* [ ] Check logic surrounding when/how pre-deploy blocks the pipeline
* [ ] Check that the monitor task roughly emulates a "cf app myapp" type API call
* [ ] Check that the kind of stats collected are useful enough
* [ ] Check that the conditions that cause the test to fail are suitable

## Who can review

Not @chrisfarms
Not @46bit 
Not @combor (although it changed so much since he was on it that it's probably ok)
